cli/src/commands/log.rs

```diff
@@ -76,23 +77,34 @@ pub(crate) fn cmd_log(
 ) -> Result<(), CommandError> {
     let workspace_command = command.workspace_helper(ui)?;
 
-    let revset_expression = {
-        let mut expression = if args.revisions.is_empty() {
-            workspace_command.parse_revset(&command.settings().default_revset())?
-        } else {
-            workspace_command.parse_union_revsets(&args.revisions)?
-        };
-        if !args.paths.is_empty() {
-            let repo_paths: Vec<_> = args
-                .paths
-                .iter()
-                .map(|path_arg| workspace_command.parse_file_path(path_arg))
-                .try_collect()?;
-            expression.intersect_with(&RevsetExpression::filter(RevsetFilterPredicate::File(
-                Some(repo_paths),
-            )));
+    let path_expression = if !args.paths.is_empty() {
+        let repo_paths: Vec<_> = args
+            .paths
+            .iter()
+            .map(|path_arg| workspace_command.parse_file_path(path_arg))
+            .try_collect()?;
+        Some(RevsetExpression::filter(RevsetFilterPredicate::File(Some(
+            repo_paths,
+        ))))
+    } else {
+        None
+    };
+
+    // if no args are passed, use default revset
+    // if only paths are passed, filter by paths
+    // if only a revset is passed, use the revset
+    // if paths and revset are passed, use an intersection of both
+    let revset_expression = match (args.revisions.is_empty(), path_expression.clone()) {
+        (true, None) => workspace_command.parse_revset(&command.settings().default_revset())?,
+        (true, Some(path_expression)) => {
+            workspace_command.attach_revset_evaluator(path_expression)?
+        }
+        (false, None) => workspace_command.parse_union_revsets(&args.revisions)?,
+        (false, Some(path_expression)) => {
+            let mut expression = workspace_command.parse_union_revsets(&args.revisions)?;
+            expression.intersect_with(&path_expression);
+            expression
```

Comment:
Perhaps, the initial set can be encoded as follows:
```rust
if args.revisions.is_empty() && args.paths.is_empty() {
    workspace_command.parse_revset(&command.settings().default_revset())?
} else if !args.revisions.is_empty() {
    workspace_command.parse_union_revsets(&args.revisions)?
} else {
    workspace_command.attach_revset_evaluator(RevsetExpression::all())?
}
```

It will be probably easier to extend when implementing "diff" predicate.
https://github.com/martinvonz/jj/issues/2933#issuecomment-1925870731

Reply:
Updated and squashed